splice_machine/src/main/java/com/splicemachine/derby/stream/utils/ExternalTableUtils.java

```diff
@@ -0,0 +1,204 @@
+            throw StandardException.newException(SQLState.INCONSISTENT_NUMBER_OF_ATTRIBUTE,
+                    tableFields.length, dataFields.length, location);
+        }
+
+        StructField[] partitionedTableFields = new StructField[tableSchema.fields().length];
+        Set<Integer> partitionColumns = new HashSet<>();
+        for (int pos : partitionColumnMap) {
+            partitionColumns.add(pos);
+        }
+        int index = 0;
+        for (int i = 0; i < tableFields.length; ++i) {
+            if (!partitionColumns.contains(i)) {
+                partitionedTableFields[index++] = tableFields[i];
+            }
+        }
+        for (int pos:partitionColumnMap) {
+            partitionedTableFields[index++] = tableFields[pos];
+        }
+
+        for (int i = 0; i < tableFields.length - partitionColumnMap.length; ++i) {
+
+            String tableFiledTypeName = partitionedTableFields[i].dataType().typeName();
+            String dataFieldTypeName = dataFields[i].dataType().typeName();
+            if (!tableFiledTypeName.equals(dataFieldTypeName)){
+                throw StandardException.newException(SQLState.INCONSISTENT_DATATYPE_ATTRIBUTES,
+                        tableFields[i].name(),
+                        tableFields[i].dataType().toString(),
+                        dataFields[i].name(),
+                        dataFields[i].dataType().toString(),location);
+            }
+        }
+    }
+
+    private static StructType getDataSchema(DataSetProcessor dsp, StructType tableSchema, int[] partitionColumnMap,
+                                     String location, String storeAs, boolean mergeSchema) throws StandardException {
+        StructType dataSchema =dsp.getExternalFileSchema(storeAs, location, mergeSchema);
+        tableSchema =  ExternalTableUtils.supportAvroDateType(tableSchema, storeAs);
+        ExternalTableUtils.checkSchema(tableSchema, dataSchema, partitionColumnMap, location);
+
+        // set partition column datatype, because the inferred type is not always correct
+        setPartitionColumnTypes(dataSchema, partitionColumnMap, tableSchema);
+        return dataSchema;
+    }
+
+    public static StructType getDataSchema(DataSetProcessor dsp, StructType tableSchema, int[] partitionColumnMap,
+                                           String location, String storeAs) throws StandardException {
+        // Infer schema from external files\
+        StructType dataSchema = null;
+        try {
+            dataSchema = getDataSchema(dsp, tableSchema, partitionColumnMap, location, storeAs, false);
+        }
+        catch (StandardException e) {
+            String sqlState = e.getSqlState();
+            if (sqlState.equals(SQLState.INCONSISTENT_NUMBER_OF_ATTRIBUTE) ||
+                    sqlState.equals(SQLState.INCONSISTENT_DATATYPE_ATTRIBUTES)) {
+                dataSchema = getDataSchema(dsp, tableSchema, partitionColumnMap, location, storeAs, true);
```

Comment:
If we hit INCONSISTENT_DATATYPE_ATTRIBUTES error, how would calling getDataSchema with mergeSchema as true resolve the error?  I thought merging schemas would just pick up more columns that exist in other parquet files, but if two columns in 2 separate parquet files have the same name but different data types, wouldn't this still fail, or is it actually smart enough to do stuff like combine an int and a long into a long?  If not, maybe it's best to let this fail on the first call to getDataSchema so we could avoid a lengthy schema inference, when we know in advance it's going to fail anyway.

Reply:
There could be schema evolution, for example dropping a column and add another column. When mergeSchema is false, it only reads schema from one file, which may have outdated metadata. If data types does not match, fall back with mergeSchema=true to get complete schema.

Reply:
@jyuanca I see.  If the user accidentally enters the wrong data type for a column, but no schema evolution took place, it may take a long time for them to get back an error message.  It would be nice if there were a user option or database property to specify not to check for schema evolution.  It may be worth creating another JIRA.

Reply:
@msirek The user should get the error when he was trying to create the table on existing data. We have an option for the user to merge schema or not. Once the table is created, we assume table and data schema matches.
 
When a user reads the table, if schema does not match when mergeSchema is false, we are pretty sure it should match if mergeSchema is true.

Reply:
@jyuanca OK, so you're saying the above code is only invoked after the external table is already created, so therefore retrying the call to getDataSchema with mergeSchema as true should always succeed?

Reply:
@msirek It should always succeed, unless the data is tampered. In that case, we will find a mismatch and throw an error.